### ISSUE TITLE
dspy application cannot be interrupted by ctrl-c

### ISSUE THREAD

**Issue by mepwang** [stanfordnlp/dspy#1686]
dspy version: 2.5.16
fastapi version: 0.115.3
os: ubuntu 24.04
python version: 3.12.3
LLM provider: deepseek

my application is a web post api in which make request to llm,.
when application waiting for response from llm, ctrl-c not work.

what should i do?

**Comment by okhat** [comment]
Thank you @mepwang ! @krypticmouse is looking into parallelism, so I imagine this _could_ be related.

@mepwang Are you using dspy.Evaluate with multi-threading? If not, I don't see why this may happen. Can you share a minimal example of this bug?

**Comment by mepwang** [comment]
> Thank you @mepwang ! @krypticmouse is looking into parallelism, so I imagine this _could_ be related.
> 
> @mepwang Are you using dspy.Evaluate with multi-threading? If not, I don't see why this may happen. Can you share a minimal example of this bug?

just plain old dspy.signature.
multithreading? maybe, the signature is called in a fastapi app.
i'll try to reproduce with minimal code.

**Comment by okhat** [comment]
Awesome, thanks @mepwang ! Looking forward to a minimal example of the issue.

**Comment by mepwang** [comment]
os: ubuntu 24.04
python version: 3.12.3

```
python -m venv env
source env/bin/activate
pip install -U pip setuptools
pip install fastapi dspy-ai uvicorn python-dotenv
```

execute:
`pip list`
result:
```
Package                   Version
------------------------- -----------
aiohappyeyeballs          2.4.3
aiohttp                   3.10.10
aiosignal                 1.3.1
alembic                   1.13.3
annotated-types           0.7.0
anyio                     4.6.2.post1
attrs                     24.2.0
backoff                   2.2.1
certifi                   2024.8.30
charset-normalizer        3.4.0
click                     8.1.7
colorlog                  6.8.2
datasets                  3.0.2
dill                      0.3.8
diskcache                 5.6.3
distro                    1.9.0
dspy                      2.5.18
dspy-ai                   2.5.18
fastapi                   0.115.4
filelock                  3.16.1
frozenlist                1.5.0
fsspec                    2024.9.0
greenlet                  3.1.1
h11                       0.14.0
httpcore                  1.0.6
httpx                     0.27.2
huggingface-hub           0.26.1
idna                      3.10
importlib_metadata        8.5.0
Jinja2                    3.1.4
jiter                     0.6.1
joblib                    1.4.2
json_repair               0.30.0
jsonschema                4.23.0
jsonschema-specifications 2024.10.1
litellm                   1.49.1
magicattr                 0.1.6
Mako                      1.3.6
MarkupSafe                3.0.2
multidict                 6.1.0
multiprocess              0.70.16
numpy                     2.1.2
openai                    1.52.2
optuna                    4.0.0
packaging                 24.1
pandas                    2.2.3
pip                       24.3.1
propcache                 0.2.0
pyarrow                   17.0.0
pydantic                  2.9.2
pydantic_core             2.23.4
python-dateutil           2.9.0.post0
python-dotenv             1.0.1
pytz                      2024.2
PyYAML                    6.0.2
referencing               0.35.1
regex                     2024.9.11
requests                  2.32.3
rpds-py                   0.20.0
setuptools                75.2.0
six                       1.16.0
sniffio                   1.3.1
SQLAlchemy                2.0.36
starlette                 0.41.2
structlog                 24.4.0
tiktoken                  0.8.0
tokenizers                0.20.1
tqdm                      4.66.5
typing_extensions         4.12.2
tzdata                    2024.2
ujson                     5.10.0
urllib3                   2.2.3
uvicorn                   0.32.0
xxhash                    3.5.0
yarl                      1.16.0
zipp                      3.20.2
```

minimal example:
```
import os
import dspy
from fastapi import FastAPI
import uvicorn

from dotenv import load_dotenv
load_dotenv()

lm = dspy.LM(
        model=os.environ.get('MODEL_PROVIDER') + '/' +
        os.environ.get('MODEL_NAME'),
        api_key=os.environ.get('OPENAI_API_KEY'),
        base_url=os.environ.get('OPENAI_API_BASE'),
        max_tokens=4096,
    )

dspy.settings.configure(lm=lm)

app = FastAPI()


@app.get('/')
async def hello():
    print('begin to predict')
    qa = dspy.Predict('question: str -> response: str')
    answer = qa(question='give me a detail tutorial about dspy framework?').response

    return {'answer': answer}

if __name__ == '__main__':
    server = uvicorn.Server(uvicorn.Config(
        app,
        host='0.0.0.0',
        port=8008,
        log_level='debug'
    ))

    server.run()
```

while serving 127.0.0.1:8008 in browser, ctrl-c not responding.
this is information in terminal
```
INFO:     Started server process [21934]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8008 (Press CTRL+C to quit)
begin to predict
^C^C^C^C^C^C^C^C^C^C^C^C^C^CINFO:     127.0.0.1:36416 - "GET / HTTP/1.1" 200 OK
```


when servs over and there is no web requst, ctrl-c works normally.

**Comment by isaacbmiller** [comment]
Allegedly will be resolved once async (#1734) merges.

Will come back after

**Comment by michael-berk** [comment]
Looks like issue is resolved.

```
dspy== 2.5.41
fastapi==0.115.5
uvicorn==0.32.1
python-dotenv==1.0.1
```
<img width="827" alt="image" src="https://github.com/user-attachments/assets/8174d10d-408c-4678-82b9-1a7415e83d94">

```
import os
import dspy
from fastapi import FastAPI
import uvicorn

from dotenv import load_dotenv
load_dotenv()

lm = dspy.LM('openai/gpt-4o-mini')
dspy.settings.configure(lm=lm)

app = FastAPI()


@app.get('/')
async def hello():
    print('begin to predict')
    qa = dspy.Predict('question: str -> response: str')
    answer = qa(question='give me a detail tutorial about dspy framework?').response

    return {'answer': answer}

if __name__ == '__main__':
    server = uvicorn.Server(uvicorn.Config(
        app,
        host='0.0.0.0',
        port=8008,
        log_level='debug'
    ))

    server.run()
```

cc @okhat
